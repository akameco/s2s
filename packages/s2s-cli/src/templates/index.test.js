// @flow
import { resolve, join } from 'path'
import rimraf from 'rimraf'
import cpFile from 'cp-file'
import stripAnsi from 'strip-ansi'
import pathExists from 'path-exists'
import * as templates from '.'

let logSpy
const fn = x => x

const fixturesPath = resolve(__dirname, '__tests__', 'fixtures')
const getEventPath = (...x) => join(fixturesPath, ...x)
const eventPath = getEventPath('copy-result.js')

beforeEach(() => {
  rimraf.sync(getEventPath('copy-result.js'))
  rimraf.sync(getEventPath('output.js'))
  logSpy = jest.spyOn(console, 'log').mockImplementation(fn)
})

afterEach(() => {
  logSpy.mockRestore()
})

test('called log with output result', () => {
  const template = { test: /copy-result.js/, input: 'copy.js' }

  templates.default(eventPath, [template], fixturesPath)
  expect(stripAnsi(logSpy.mock.calls[0][0])).toMatchSnapshot()
})

test('already exist', () => {
  const template = { test: /copy-result.js/, input: 'copy.js' }

  templates.default(eventPath, [template], fixturesPath)
  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy.mock.calls).toHaveLength(1)
})

test('handle copy error', () => {
  const template = { test: /copy-result.js/, input: 'not-found.js' }

  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy.mock.calls[0][0]).toMatch('no such file or directory')
})

test('handle nomal error', () => {
  const template = { test: /copy-result.js/, input: 'not-found.js' }

  const cpFIleSpy = jest.spyOn(cpFile, 'sync').mockImplementation(() => {
    throw new Error('hello')
  })
  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy.mock.calls[0][0]).toMatch('hello')
  cpFIleSpy.mockRestore()
})

test('template not found', () => {
  const template = { test: /copy-result.js/, input: 'not-found.js' }

  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy).toHaveBeenCalled()
  expect(logSpy.mock.calls[0][0]).toMatch('no such file or directory')
})

test('show output path when add output option', () => {
  const template = {
    test: /copy-result.js/,
    input: 'copy.js',
    output: 'output.js',
  }

  templates.default(eventPath, [template], fixturesPath)
  expect(pathExists.sync(getEventPath('output.js'))).toBe(true)
  expect(stripAnsi(logSpy.mock.calls[0][0])).toMatchSnapshot()
})

test('test regex not match', () => {
  const template = { test: /not-found.js/, input: 'not-found.js' }

  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy).not.toHaveBeenCalled()
})

test('no result when templates = []', () => {
  templates.default(eventPath, [])
  expect(logSpy).not.toHaveBeenCalled()
})

test('no result when templates = undefined', () => {
  templates.default(eventPath)
  expect(logSpy).not.toHaveBeenCalled()
})

test('handlePluginsのPlugin.testオプションはglobを判定できる', () => {
  const template = { test: '**/*.ts', input: 'not-found.js' }

  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy).not.toHaveBeenCalled()
})

test('handlePluginsのPlugin.testオプションはglobの配列を判定できる', () => {
  const template = { test: ['**/*.js', '!copy'], input: 'copy.js' }

  templates.default(eventPath, [template], fixturesPath)
  expect(logSpy).toHaveBeenCalled()
})
