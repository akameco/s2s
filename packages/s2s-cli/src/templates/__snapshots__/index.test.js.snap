// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`called log with output result 1`] = `"TEMPLATE packages/s2s-cli/src/templates/__tests__/fixtures/copy.js → packages/s2s-cli/src/templates/__tests__/fixtures/copy-result.js"`;

exports[`esj 1`] = `
"// copy

class fixtures extends React.Component {
  render() {
    return <div>{copy-result}</div>
  }
}
"
`;

exports[`show output path when add output option 1`] = `"TEMPLATE packages/s2s-cli/src/templates/__tests__/fixtures/copy.js → packages/s2s-cli/src/templates/__tests__/fixtures/output.js"`;
